File: src/main/java/gargoyle/ct/mutex/FileMutex.java
package gargoyle.ct.mutex;

import gargoyle.ct.log.Log;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;

public final class FileMutex implements CTMutex {

    private final String           mutex;
    private       File             file;
    private       RandomAccessFile randomAccessFile;
    private       FileLock         fileLock;
    private       FileChannel      channel;

    public FileMutex(String mutex) {
        this.mutex = mutex;
    }

    @Override
    public synchronized boolean acquire() {
        try {
            file = new File(mutex);
            randomAccessFile = new RandomAccessFile(file, "rw");
            channel = randomAccessFile.getChannel();
            fileLock = channel.tryLock();
            if (fileLock != null) {
                Runtime.getRuntime().addShutdownHook(new Thread(this::release));
                return true;
            }
        } catch (IOException ex) {
            // log.error("Unable to create and/or lock file: " + lockFile, e);
            Log.error(ex, MSG_MUTEX_ERROR);
        }
        return false;
    }

    @Override
    public synchronized void release() {
        if (file != null && file.exists() && fileLock != null && randomAccessFile != null) {
            try {
                fileLock.release();
                randomAccessFile.close();
                if (!file.delete()) {
                    throw new IOException(MSG_MUTEX_ERROR);
                }
                fileLock = null;
                randomAccessFile = null;
                file = null;
                channel = null;
            } catch (IOException ex) {
                Log.error(ex, MSG_MUTEX_ERROR);
            }
        }
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/mutex/impl/FileMutex.kt
package gargoyle.ct.util.mutex.impl

import gargoyle.ct.util.log.Log
import gargoyle.ct.util.mutex.CTMutex
import java.io.File
import java.io.IOException
import java.io.RandomAccessFile
import java.nio.channels.FileChannel
import java.nio.channels.FileLock
import java.nio.channels.OverlappingFileLockException

class FileMutex(private val mutex: String) : CTMutex {
    private lateinit var channel: FileChannel
    private lateinit var file: File
    private lateinit var fileLock: FileLock
    private lateinit var randomAccessFile: RandomAccessFile

    @Synchronized
    override fun acquire(): Boolean {
        if (::file.isInitialized && file.exists()) {
            return false
        }
        try {
            file = File(
                File(
                    System.getProperty(
                        "java.io.tmpdir",
                        System.getProperty("user.home", System.getProperty("user.dir", "."))
                    )
                ),
                "$mutex.lock"
            )
            randomAccessFile = RandomAccessFile(file, "rw")
            channel = randomAccessFile.channel
            channel.tryLock()?.also { fileLock = it }
            if (::fileLock.isInitialized) {
                Runtime.getRuntime().addShutdownHook(Thread { release() })
                file.deleteOnExit()
                return true
            }
        } catch (ex: IOException) {
            Log.error(ex, "file mutex error")
            return false
        } catch (ex: OverlappingFileLockException) {
            return false
        }
        return false
    }

    @Synchronized
    override fun release() {
        if (::file.isInitialized && file.exists()) {
            try {
                fileLock.release()
                randomAccessFile.close()
                if (!file.delete()) throw IOException("file mutex error")
//                fileLock = null
//                randomAccessFile = null
//                file = null
//                channel = null
            } catch (ex: IOException) {
                Log.error(ex, "file mutex error")
            }
        }
    }

}
